auth-server/src/services/auth.ts

```diff
@@ -138,7 +165,71 @@ export default class AuthService {
 		await User
 			.query()
 			.patch({ name: newName })
+			.findById(userId)
+	}
+
+	public async VerifyEmail(token: string) {
+		const verifyToken = await VerifyToken
+			.query()
+			.where('token', token)
+			.first()
+
+		if (!verifyToken) {
+			throw new Error('Verify token not found')
+		}
+
+		if (!verifyToken.valid) {
+			throw new Error('Verify token not valid')
+		}
+
+		await User
+			.query()
+			.patch({ email_verified: true })
+			.findById(verifyToken.id)
+
+		await VerifyToken
+			.query()
+			.patch({ valid: false })
+			.findById(verifyToken.id)
+	}
+
+	public async ChangeEmail(token: string, email: string) {
+		// verify a token symmetric - synchronous
+		const decoded = jwt.verify(token, publicKey)
+
+		if (isNaN(decoded.sub)) {
+			throw new Error('Invalid user id')
+		}
+
+		const user = await User
+			.query()
+			.where('id', Number(decoded.sub))
+			.first()
+
+		if (!user) {
+			throw new Error('User not found')
+		}
+
```

Comment:
We must to invalidate any other token that could have been created (and not used) by this user, so that we only ever have 1 valid token per user otherwise we'll always send emails to the first email (which might be wrong).

Reply:
invalidating other tokens